catfs/db/database_badger.go

```diff
@@ -56,17 +55,95 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
 				return
 			}
 
-			if err := db.RunValueLogGC(0.5); err != nil {
-				if err != badger.ErrNoRewrite {
-					log.WithError(err).Warnf("badger gc failed")
-				}
+			err := bdb.runGC()
+			if err != nil {
+				log.WithError(err).Error("badger GC failed")
 			}
 		}
 	}()
 
 	return bdb, nil
 }
 
+func (bdb *BadgerDatabase) runGC() error {
+	opts := bdb.db.Opts()
+	// The logic is quite convoluted:
+	// It is not enough to ask for GC, badger DB needs to run some statistic
+	// after last modification in DB (about a minute), if we don't wait
+	// the GC call will be waisted in we might claim that there is nothing to GC.
+	// Also to do GC correctly we need to Close DB before calling GC
+	// (yep, badger is crazy in this regard).
+	// Then we reopen DB and mark cleanable with GC,
+	// then on Close() true GC happens (HDD space reclaimed),
+	// and we need to reopen again.
+	// Consequently we have a timer (to give time for stats update) and we needGC flag
+	// which indicates that some updates were done on DB so we do not trash
+	// underlying FS.
+	// TODO: to do it really correct we need a queue which holds last DB modification
+	// times. But it all become very convoluted since other processes mess with DB
+	// Or repinner always does something in background and updates DB.
+	if time.Now().Before(bdb.gcNeededAt) || !bdb.needsGC {
+		// we are waiting for timer but also for updates
+		// if there are no updated in DB, timer does not kick in
+		log.Debugf("badger DB in %s stats are not updated yet, no need for GC", opts.Dir)
+		return nil
+	}
+	bdb.mu.Lock()
+	defer bdb.mu.Unlock()
+	defer func() {
+		bdb.gcNeededAt = time.Now()
+		bdb.needsGC = false
+	}()
+	var errGC error
+	var beforeGC uint64 = 0
+	var total uint64 = 0
+	var err error
+	log.Infof("Performin GC for badger DB in %s", opts.Dir)
```

Comment:
typo: "Performing".

Reply:
fixed, but there is a new patch which makes it obsolete.